python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -993,12 +1012,63 @@ def build_asset_outs(asset_outs: Mapping[str, AssetOut]) -> Mapping[AssetKey, Tu
     return outs_by_asset_key
 
 
-def _make_asset_keys(deps: Optional[Union[Set[AssetKey], Set[str]]]) -> Optional[Set[AssetKey]]:
-    """Convert all str items to AssetKey in the set."""
+def _type_check_deps_and_non_argument_deps(
+    deps: Optional[Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]]],
+    non_argument_deps: Optional[Union[Set[AssetKey], Set[str]]],
+):
+    """Helper function for managing deps and non_argument_deps while non_argument_deps is still an accepted parameter.
+    Ensures:
+    1. only one of deps and non_argument_deps is provided.
+    2. multi assets AssetsDefinition is not passed to deps.
+    3. deprecation warning is fired for non_argument_deps.
+    """
+    if non_argument_deps is not None and deps is not None:
+        raise DagsterInvalidDefinitionError(
+            "Cannot specify both deps and non_argument_deps to @asset. Use only deps instead."
+        )
+
+    upstream_asset_deps: Optional[
+        Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]]
+    ] = None
+    if deps is not None:
+        check.list_param(deps, "deps", of_type=(CoercibleToAssetKey, AssetsDefinition, SourceAsset))
```

Comment:
adding this check in makes everything fail with ` TypeError: Subscripted generics cannot be used with class and instance checks` errors. I think this is because `CoercibleToAssetKey` contains `List[str]`. I'm not sure what to do about this though

Reply:
i think it is understandable that you'll need to write some manual runtime checks instead of  leaning on `check.list_` , should combine reasonably well with the other checks you are doing

Reply:
yup - ended up modifying into what it is right now. The error message is pretty minimal - I could try to break out the `CacheableAssetsDefinition` to a separate error message, but need to figure out what the error message would be first